File: compilers/rjs/transpiler/parser/src/ast/statement.rs
use crate::ast::expression::{Expression, Identifier, Pattern};
use crate::ast::declaration::VariableDeclaration;
use crate::ast::StatementListItem;
use serde::{Deserialize, Serialize};

use super::declaration::FunctionDeclaration;

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize,)]
pub struct IfStatement<'a> {
    pub test: Expression<'a>,
    pub conseq: Box<Statement<'a>>,
    pub alter: Option<Box<Statement<'a>>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct BlockStatement<'a> {
    pub body: Vec<StatementListItem<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct SwitchStatement<'a> {
    pub deicriminant: Expression<'a>,
    pub cases: Vec<SwitchCase<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct SwitchCase<'a> {
    pub test: Option<Expression<'a>>,
    pub conseq: Vec<StatementListItem<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize )]
pub struct ContinueStatement<'a> {
    pub label: Option<Identifier<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct BreakStatement<'a> {
    pub label: Option<Identifier<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ReturnStatement<'a> {
    pub argument: Option<Expression<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct LabeledStatement<'a> {
    pub label: Identifier<'a>,
    pub body: LabeldItem<'a>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(untagged)]
pub enum LabeldItem<'a> {
    Stmt(Box<Statement<'a>>),
    FunDeclar(FunctionDeclaration<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct WhileStatement<'a> {
    pub test: Expression<'a>,
    pub body: Box<Statement<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct DoWhileStatement<'a> {
    pub test: Expression<'a>,
    pub body: Box<Statement<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct TryStatement<'a> {
    pub block: BlockStatement<'a>,
    pub handler: Option<CatchClause<'a>>,
    pub finalizer: Option<BlockStatement<'a>>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ThrowStatement<'a> {
    pub argument: Expression<'a>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct CatchClause<'a> {
    pub param: Option<Pattern<'a>>,
    pub block: BlockStatement<'a>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ExpressionStatement<'a> {
    pub expr: Expression<'a>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct DebuggerStatement;
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct EmptyStatement;
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ForStatement<'a> {
    pub init: Option<ForInit<'a>>,
    pub test: Option<Expression<'a>>,
    pub update: Option<Expression<'a>>,
    pub body: Box<Statement<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ForOfStatement<'a> {
    pub left: ForInit<'a>,
    pub right: Expression<'a>,
    pub is_async: bool,
    pub body: Box<Statement<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ForInStatement<'a> {
    pub left: ForInit<'a>,
    pub right: Expression<'a>,
    pub body: Box<Statement<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ForInit<'a> {
    Expr(Expression<'a>),
    Var(VariableDeclaration<'a>),
    Pat(Pattern<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ForRelateStatement<'a> {
    For(ForStatement<'a>),
    ForIn(ForInStatement<'a>),
    ForOf(ForOfStatement<'a>)
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(tag="type")]
pub enum Statement<'a> {
    #[serde(rename="IfStatement")]
    IfStmt(IfStatement<'a>),
    #[serde(rename="BlockStatement")]
    BlockStmt(BlockStatement<'a>),
    #[serde(rename="SwitchStatement")]
    SwitchStmt(SwitchStatement<'a>),
    #[serde(rename="ContinueStatement")]
    ContinueStmt(ContinueStatement<'a>),
    #[serde(rename="BreakStatement")]
    BreakStmt(BreakStatement<'a>),
    #[serde(rename="ReturnStatement")]
    ReturnStmt(ReturnStatement<'a>),
    #[serde(rename="LabeledStatement")]
    LabeledStmt(LabeledStatement<'a>),
    #[serde(rename="WhileStatement")]
    WhileStmt(WhileStatement<'a>),
    #[serde(rename="DoWhileStatement")]
    DoWhileStmt(DoWhileStatement<'a>),
    #[serde(rename="TryStatement")]
    TryStmt(TryStatement<'a>),
    #[serde(rename="DebuggerStatement")]
    DebuggerStmt(DebuggerStatement),
    #[serde(rename="EmptyStatement")]
    EmptyStmt(EmptyStatement),
    #[serde(rename="ExpressionStatement")]
    ExprStmt(ExpressionStatement<'a>),
    #[serde(rename="ForStatement")]
    ForStmt(ForStatement<'a>),
    #[serde(rename="ForInStatement")]
    ForInStmt(ForInStatement<'a>),
    #[serde(rename="ForOfStatement")]
    ForOfStmt(ForOfStatement<'a>),
}

File: web-infras/common/src/ast/statement.ts
import { ModuleItem } from "@/src/ast/base";
import { SyntaxKinds } from "@/src/kind";
import { StatementListItem } from "@/src/ast/program";
import { Expression, Identifier } from "@/src/ast/expression";
import { VariableDeclaration, FunctionDeclaration } from "@/src/ast/declaration";
import { Pattern } from "@/src/ast/pattern";

export interface IfStatement extends ModuleItem {
  kind: SyntaxKinds.IfStatement;
  test: Expression;
  conseqence: Statement;
  alternative: Statement | null;
}
export interface BlockStatement extends ModuleItem {
  kind: SyntaxKinds.BlockStatement;
  body: Array<StatementListItem>;
}
export interface SwitchStatement extends ModuleItem {
  kind: SyntaxKinds.SwitchStatement;
  discriminant: Expression;
  cases: Array<SwitchCase>;
}
export interface SwitchCase extends ModuleItem {
  kind: SyntaxKinds.SwitchCase;
  test: Expression | null;
  consequence: Array<StatementListItem>;
}
export interface ContinueStatement extends ModuleItem {
  kind: SyntaxKinds.ContinueStatement;
  label: Identifier | null;
}
export interface BreakStatement extends ModuleItem {
  kind: SyntaxKinds.BreakStatement;
  label: Identifier | null;
}
export interface ReturnStatement extends ModuleItem {
  kind: SyntaxKinds.ReturnStatement;
  argu: Expression | null;
}
export interface LabeledStatement extends ModuleItem {
  kind: SyntaxKinds.LabeledStatement;
  label: Identifier;
  body: Statement | FunctionDeclaration;
}
export interface WhileStatement extends ModuleItem {
  kind: SyntaxKinds.WhileStatement;
  test: Expression;
  body: Statement;
}
export interface DoWhileStatement extends ModuleItem {
  kind: SyntaxKinds.DoWhileStatement;
  test: Expression;
  body: Statement;
}
export interface TryStatement extends ModuleItem {
  kind: SyntaxKinds.TryStatement;
  block: BlockStatement;
  handler: CatchClause | null;
  finalizer: BlockStatement | null;
}
export interface CatchClause extends ModuleItem {
  kind: SyntaxKinds.CatchClause;
  param: Pattern | null;
  body: BlockStatement;
}
export interface ThrowStatement extends ModuleItem {
  kind: SyntaxKinds.ThrowStatement;
  argu: Expression;
}
export interface WithStatement extends ModuleItem {
  kind: SyntaxKinds.WithStatement;
  object: Expression;
  body: Statement;
}
export interface DebuggerStatement extends ModuleItem {
  kind: SyntaxKinds.DebuggerStatement;
}
export interface EmptyStatement extends ModuleItem {
  kind: SyntaxKinds.EmptyStatement;
}
export interface ForStatement extends ModuleItem {
  kind: SyntaxKinds.ForStatement;
  init: Expression | VariableDeclaration | null;
  test: Expression | null;
  update: Expression | null;
  body: Statement;
}
export interface ForOfStatement extends ModuleItem {
  kind: SyntaxKinds.ForOfStatement;
  left: Expression | VariableDeclaration;
  right: Expression;
  await: boolean;
  body: Statement;
}
// TODO. better Type
// type ForOfInStatementLeft = VariableDeclaration | Pattern | TSAsExpression | TSTypeAssertionExpression;
export interface ForInStatement extends ModuleItem {
  kind: SyntaxKinds.ForInStatement;
  left: Expression | VariableDeclaration;
  right: Expression;
  body: Statement;
}

export interface ExpressionStatement extends ModuleItem {
  kind: SyntaxKinds.ExpressionStatement;
  expr: Expression;
}

export type Statement =
  | IfStatement
  | BlockStatement
  | SwitchStatement
  | ForInStatement
  | ForOfStatement
  | ForStatement
  | BreakStatement
  | ContinueStatement
  | ReturnStatement
  | LabeledStatement
  | WhileStatement
  | DoWhileStatement
  | TryStatement
  | ThrowStatement
  | WithStatement
  | DebuggerStatement
  | EmptyStatement
  | ExpressionStatement
  | VariableDeclaration /** when is `var` */;
